### PR TITLE
bundle custom lib in docker image

### LINKED ISSUE #85: bundling local libraries when building Docker image
Users may want to install packages that are not available publicly on PyPI. A typical use case is an internal library stored in some git repository or a private registry.

For example, assume the local environment is properly configured to authenticate with a private git repository, someone can easily install that package with a `requirements.txt`:

```
git+https://path.to/repo
```

However, in some cases, `soopervisor export` might not run locally, but in a CI/CD server; hence installing from private repositories won't work unless the CI/CD server is properly authenticated to get the repository (`https://path.to/repo`).

This also applies to Ploomber Cloud users since the Docker image is built remotely. In Ploomber Cloud's case is even more desirable to have this solution since users will prefer not to disclose their credentials.

The solution is for the Docker image to be able to pick up packages from a local directory. For example, a user might have the following layout:

```
lib/
  package_a/
  package_b/

tasks/
  load.py
  clean.py
  fit.py

pipeline.yaml
```

The Dockerfile should configure the PYTHONPATH so that it includes the `lib/` directory; hence when starting Python, `import package_a` and `import package_b` will work.

To get `package_a` under `lib`, users can use `pip install`, but we should do some research to find out how to do it, these two look promising:

```
  --root <dir>                Install everything relative to this alternate root
                              directory.
  --prefix <dir>              Installation prefix where lib, bin and other top-level
                              folders are placed
```

The final thing to take into account is that for packages that are not pure-Python (e.g. the ones that use C extensions), if the local OS is different than the OS where the pipeline will be executed; we might run into issues; but I think we can ignore that for now.

cc @idomic @Wxl19980214

### PR DESCRIPTION
**Pull request by Wxl19980214**
## Describe your changes
Add path check. If Path(lib/') exist, add a PYTHONPATH in docerfile to look for packages under lib.

## Issue ticket number and link
Closes #85 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added thorough tests (when necessary).
- [x] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

### CODE REVIEW

**Comment by edublancas** [comment]
please write an automated test, see this for details: https://github.com/ploomber/soopervisor/issues/88

**Comment by Wxl19980214** [comment]
@edublancas  Yes. I have tested in this way. This is my project structure:
project/
          lib/ 
               package_a.py (contains function printTest() )
          fit.py  (from lib.package_a import printTest )
          ....

And I use sooperviosr add and export to build the image. It looks like there's no error when I run the image.

**Comment by idomic** [comment]
@Wxl19980214 please add a few tests.

**Comment by Wxl19980214** [comment]
@edublancas  @idomic I am having troubling writing test: Here is what I have:
<pre>
@pytest.mark.parametrize('CLASS_', [
    AWSBatchExporter,
    AirflowExporter,
    ArgoWorkflowsExporter,
    KubeflowExporter,
],
                         ids=[
                             'batch',
                             'airflow',
                             'argo',
                             'kubeflow',
                         ])
def test_docker_local_lib_import(
    mock_docker_and_batch,
    tmp_sample_project,
    no_sys_modules_cache,
    skip_repo_validation,
    capsys,
    CLASS_,
):

    exporter = CLASS_.new(path_to_config='soopervisor.yaml', env_name='serve')

    exporter.add()
    exporter.export(mode='force', skip_tests=True)

    client = docker.from_env()
    client.containers.run('python', 'import package_a')
    
    captured = capsys.readouterr()

    print(captured.out)


</pre>

I have added the lib/ directory with package_a under asset/sample_project. However, it seems like the docker.from_env() is mocked so I cannot get real result. Any ideas?

**Comment by Wxl19980214** [comment]
Also for some reason, i cannot recreate what I had last week. After soopervisor export, what command do we use to run the image? docker run example:latest? I have tried and it looks like not working

**Comment by idomic** [comment]
That's right. After the export command you should have the image tagged locally and you can simply use `docker run tag -it` to run it and debug. You can also list the available images to see if it's not there for some reason.

**Comment by idomic** [comment]
It should be something like this `docker run image_name:tag_name`

**Comment by Wxl19980214** [comment]
docer run example:latest? 
It just returns nothing and not actually running the image.

**Comment by idomic** [comment]
Issue solved.

**Comment by Wxl19980214** [comment]
@edublancas I think he meant local test are done. I am still figuring out a way to write automated tests.

**Comment by edublancas** [comment]
ok cool

File: tests/test_export.py [hunk 1]
```diff
@@ -210,3 +212,49 @@ def test_validates_repository(mock_docker_sample_project, tmp_sample_project,
...
+    CLASS_,
+):
+
+    text = "serve: {backend: kubeflow, repository: null}"
+
+    exporter = CLASS_.new(path_to_config='soopervisor.yaml', env_name='serve')
+
+    Path('soopervisor.yaml').write_text(text)
+
+    exporter.add()
```
Comment (edublancas) [inline tests/test_export.py:236]: I see what you mentioned. this line always overwrites soopervisor.yaml

try something like this:

```python

exporter.add()

# load soopervisor.yaml using the yaml library

# set repository to null

exporter.export(mode='incremental', skip_tests=True)
```

**Review by edublancas — CHANGES_REQUESTED** [review]

File: tests/environment.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+dependencies:
+- a
+- b
+- pip: 1
```
Comment (edublancas) [inline tests/environment.yml:4]: what is this file for?

**Review by edublancas — CHANGES_REQUESTED** [review]
can you add a section here: https://github.com/ploomber/soopervisor/blob/master/doc/user-guide/build-process.rst

explaining how this new feature works?

**Comment by edublancas** [comment]
is this ready for review?

**Comment by Wxl19980214** [comment]
@edublancas Yes. I couldn't find where the changes were requested. I thought I resolved them all.

**Comment by idomic** [comment]
@Wxl19980214 Did you do some manual tests as well?

**Comment by Wxl19980214** [comment]
@idomic Yes I have tried with argo and kubernate. I can go ahead and check other backend as well

**Comment by idomic** [comment]
Please test from main with another backend let's say batch

**Comment by Wxl19980214** [comment]
Just tested. There is no issue.

**Pull request merged**